components/brave_shields/browser/tracking_protection_service.cc

```diff
@@ -251,57 +255,69 @@ bool TrackingProtectionService::ShouldStartRequest(
   return false;
 }
 
-void TrackingProtectionService::OnDATFileDataReady() {
-  if (buffer_.empty()) {
+void TrackingProtectionService::OnGetDATFileData(GetDATFileDataResult result) {
+  if (result.second.empty()) {
     LOG(ERROR) << "Could not obtain tracking protection data";
     return;
   }
-  tracking_protection_client_.reset(new CTPParser());
-  if (!tracking_protection_client_->deserialize(
-          reinterpret_cast<char*>(&buffer_.front()))) {
-    tracking_protection_client_.reset();
+  if (!result.first.get()) {
     LOG(ERROR) << "Failed to deserialize tracking protection data";
     return;
   }
+
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::IO},
+      base::BindOnce(&TrackingProtectionService::UpdateTrackingProtectionClient,
+                     weak_factory_io_thread_.GetWeakPtr(),
+                     std::move(result.first),
+                     std::move(result.second)));
+}
+
+void TrackingProtectionService::UpdateTrackingProtectionClient(
+    std::unique_ptr<CTPParser> tracking_protection_client,
+    brave_component_updater::DATFileDataBuffer buffer) {
```

Comment:
DCHECK_CURRENTLY_ON(BrowserThread::IO); just in case someone uses it from non IO thread